### PR TITLE
Implement minOptional, maxOptional, minByOptional and maxByOptional o…

### LINKED ISSUE #122: Implement minOptional, maxOptional, minByOptional and maxByOptional on RichIterable.
Implement min/max methods that returns Optional on RichIterable. The method signatures would look like below.

`Optional<T> minOptional(Comparator<? super T> comparator)`
`Optional<T> maxOptional(Comparator<? super T> comparator)`
`Optional<T> minOptional()`
`Optional<T> maxOptional()`
`<V extends Comparable<? super V>> Optional<T> minByOptional(Function<? super T, ? extends V> function)`
`<V extends Comparable<? super V>> Optional<T> maxByOptional(Function<? super T, ? extends V> function)`

### PR DESCRIPTION
**Pull request by PascalSchumacher**
…n RichIterable (closes #122)

### CODE REVIEW

**Comment by PascalSchumacher** [comment]
Note that #122 suggested `minByOptional` and `maxByOptional`, but I added `minOptionalBy` and `maxOptionalBy` instead.

**Comment by motlin** [comment]
Interesting point about the name. We have a method named detectWithOptional. I think to be consistent, it would have to be called detectOptionalWith. I'd ask the whole group to chime in about the name. Once it's in, it's hard to change.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/RichIterable.java [hunk 1]
```diff
@@ -1562,6 +1562,38 @@ default T getOnly()
...
+     * @throws NullPointerException if the minimum element is null
+     * @since 8.2
+     */
+    default Optional<T> minOptional(Comparator<? super T> comparator)
+    {
+        if (this.isEmpty())
+        {
+            return Optional.empty();
+        }
+        return Optional.of(min(comparator));
```
Comment (nikhilnanivadekar) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/RichIterable.java]: Please qualify with `this`. this.min()

**Comment by nikhilnanivadekar** [comment]
We should be consistent with detectWithOptional() i.e. minByOptional(). Our users know the *By() methods take a function. While looking at auto-complete option they'll see the methods in this order:
min(), minBy(), minOptional() and minByOptional(). 
The above looks symmetrical as compared to:
min(), minBy(), minOptional() and minOptionalBy(). 
So, I vote for minByOptional().

**Comment by PascalSchumacher** [comment]
I missed `detectWithOptional`. I agree it's important to be consistent.

Just tell me the final decision and I will happily update the pull request.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/RichIterable.java (hunk 1 above)
Comment (PascalSchumacher) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/RichIterable.java]: Sorry, to get this wrong again. :(

**Comment by motlin** [comment]
I think I also prefer minByOptional/maxByOptional. I think the point @nikhilnanivadekar made about them appearing alphabetically when you press ctrl+space is a good one. Because in English I'd probably say "optionally return the min by ..." but we don't put the word optional at the beginning because it wouldn't help much with auto-complete. Want to jump in on the naming conversation @donraab?

Once we figure out naming, the PR looks great - ready to go. Thanks again!

**Comment by PascalSchumacher** [comment]
Thanks for the feedback! I have changed the method names to `minByOptional` and `maxByOptional`.

**Pull request merged**